### PR TITLE
JENKINS-39916 update config-file-provider plugin

### PR DESCRIPTION
**Pull request by imod**
https://issues.jenkins-ci.org/browse/JENKINS-39916

The new version of the config-file-provider supports configfiles on folder levels (also nested folders). This PR adds support for the newest version of the config-file-provider.

It updates dependency to the latest (still beta) version of the config-file-provider plugin. The beta version of the config-file-provider plugin is available in the experimental update center.

Please have a look at this PR and let me know it is OK - I will prepare the release of the config-file-provider plugin soon. (...there are multiple plugins I have to update)

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java [hunk 1]
```diff
@@ -517,7 +518,16 @@ private String setupGlobalSettingFile() throws IOException, InterruptedException
...
      */
     private void settingsFromConfig(String settingsConfigId, FilePath settingsFile) throws AbortException {
-        Config c = Config.getByIdOrNull(settingsConfigId);
+
+        Config c = null;
+        Executor executor = Executor.currentExecutor();
+        if (executor != null) {
+            Queue.Executable currentExecutable = executor.getCurrentExecutable();
+            if (currentExecutable != null) {
+                c = ConfigFiles.getByIdOrNull((Run<?, ?>) currentExecutable, settingsConfigId);
```
Comment (imod) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: @alvarolobato this part I'm not sure it works in this case (it does works as a workaround in other plugins I have done the migration) - but unfortunate this plugin does not have any test cases so I can not easily check it.

the best would be to have access to the build/run at this location, but I have no idea how to get it from `StepContext`.

File: src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java [hunk 2]
```diff
@@ -564,7 +574,16 @@ private void settingsFromConfig(String settingsConfigId, FilePath settingsFile)
...
      */
     private void globalSettingsFromConfig(String globalSettingsConfigId, FilePath globalSettingsFile) throws AbortException {
-        Config c = Config.getByIdOrNull(globalSettingsConfigId);
+
+        Config c = null;
+        Executor executor = Executor.currentExecutor();
+        if (executor != null) {
+            Queue.Executable currentExecutable = executor.getCurrentExecutable();
+            if (currentExecutable != null) {
+                c = ConfigFiles.getByIdOrNull((Run<?, ?>) currentExecutable, globalSettingsConfigId);
```
Comment (imod) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: same thing ^

File: src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java (hunk 1 above)
Comment (imod) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: I just checked, this unfortunate, this does not work @jglick do you have any hint how I can access the Run/Build from a `StepContext`? I need to to be able to identify the context where to look for config-files (e.g. in witch folder)

**Comment by cyrille-leclerc** [comment]
@imod I have reported a problem and proposed and improvement of the exception message on https://github.com/imod/pipeline-maven-plugin/pull/1

File: src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: `get`

Comment (imod) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: @jglick not sure whether your answer is finished or you just did hit the "Comment" button to early - ...I don't understand what you'r saying

Comment (imod) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: ...or maybe I do: in PR #11  you make `Run<?, ?> build` directly accessible - great, thanks!

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: It was already directly accessible. I just changed the manner of its injection. Not following.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: AFAICT this whole block could be replaced by

```java
Config c = ConfigFiles.getByIdOrNull(build, settingsConfigId);
```

No dependency on #11.

Comment (cyrille-leclerc) [inline src/main/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepExecution.java]: @jglick thanks, I proposed a fix on a PR on this PR: https://github.com/imod/pipeline-maven-plugin/pull/1

**Comment by cyrille-leclerc** [comment]
See fix on a PR on this PR: https://github.com/imod/pipeline-maven-plugin/pull/1

**Comment by alvarolobato** [comment]
@imod this PR is now conflicted with the last merges, I've created a PR to your repo, fixing it. 
https://github.com/imod/pipeline-maven-plugin/pull/2

**Comment by cyrille-leclerc** [comment]
Thanks @alvarolobato !

**Comment by cyrille-leclerc** [comment]
@imod when do you think that we will be able to merge this PR? 

I would need it as I am working on "Support Maven Settings Provider and Maven Global Settings Provider for pipeline and make it work with `withMaven(){}`"

See
* [JENKINS-40665 Maven SettingsProvider and GlobalSettingsProvider should support Pipeline jobs](https://issues.jenkins-ci.org/browse/JENKINS-40665)
* [JENKINS-39407 Use system default Maven, Maven Settings if not specified](https://issues.jenkins-ci.org/browse/JENKINS-39407)
* PR https://github.com/jenkinsci/jenkins/pull/2678
* PR https://github.com/jenkinsci/pipeline-maven-plugin/pull/14
* PR https://github.com/jenkinsci/config-file-provider-plugin/pull/24

**Comment by imod** [comment]
@cyrille-leclerc i just released 2.15.1 of config-file-provider, so everything should be up to date now.

I hope the maintainers of the depending plugins followup with releases soon:

* this PR
* https://github.com/jenkinsci/job-dsl-plugin/pull/951
* https://github.com/jenkinsci/openstack-cloud-plugin/pull/121
* https://github.com/jenkinsci/ivy-plugin/pull/23

**Comment by cyrille-leclerc** [comment]
We should check https://github.com/jenkinsci/pipeline-maven-plugin/pull/11 when merging

**Comment by cyrille-leclerc** [comment]
👍  Successfully tested this snapshot version of pipeline-maven-plugin with config-file-provider:2.15.1

**Comment by imod** [comment]
@cyrille-leclerc many thanks!

**Pull request merged**